### PR TITLE
Add `cuda::mr::cuda_memory_resource`

### LINKED ISSUE #1512: [FEA]: Implement a memory_resource using `cudaMalloc` and `cudaFree`
In order to move our `<memory_resource>` plans forward we want to implement a memory resource that utilizes `cudaMalloc` and `cudaFree` to allocate memory.

That memory_resource should live in `<cuda/memory_resource>` and at least satisfy the `cuda::mr::resource_with<cuda::mr::device_accessible>`  requirement.

Poll for the audience. Should it also satisfy `cuda::mr::async_resource_with<cuda::mr::device_accessible>`?

### PR DESCRIPTION
**Pull request by miscco**
This introduces a new memory resource that allocates device memory through `cudaMalloc` / `cudaFree`

Fixes #1512

### CODE REVIEW

**Comment by miscco** [comment]
@harrism you might be interested in this

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated
+   * @param __bytes The size in bytes of the allocation.
+   */
+  void deallocate(void* __ptr, size_t) const
+  {
+    const ::cudaError_t __status = ::cudaFree(__ptr);
+    _LIBCUDACXX_ASSERT(__status == cudaSuccess, "cuda_memory_resource::deallocate failed");
```
Comment (miscco) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:84]: I have been super wonky here

Do we also want a way to throw exceptions?

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda_runtime_api.h>
```
Comment (jrhemstad) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:24]: This include should be guarded.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  /**
+   * @brief Allocate device memory of size at least \p __bytes.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation. Is ignored!
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* allocate(const size_t __bytes, size_t) const
+  {
+    return allocate(__bytes);
+  }
```
Comment (jrhemstad) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:50]: cudaMalloc is guaranteed to return memory that is aligned to at least 256B. Instead of ignoring the alignment, we should check that it is <= 256, otherwise throw.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 1 above)
Comment (jrhemstad) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:84]: An assert is pretty much the best we can do in a deallocate.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 3 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:50]: I went with a simple assert until we are sure what we want to do regarding exceptions

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 2 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:24]: We decided that everything inside plain `cuda` can access `<cuda_runtime_api>` unconditionally, which is what we have done in `<cuda/stream_ref>`

Comment (jrhemstad) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:24]: You only ever need to include `cuda_runtime_api.h` when compiling with a host-only compiler.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 3 above)
Comment (jrhemstad) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:50]: What remains to be decided about exceptions? We're already throwing exceptions in the `stream_ref` code: https://github.com/NVIDIA/cccl/blob/d88fbd7341647533128f2bbe65cea9f22a874caf/libcudacxx/include/cuda/stream_ref#L139-L154

Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h]: Actually only me needing to write a proper wrapper for that

File: libcudacxx/include/cuda/__memory_resource/cuda_pinned_memory_resource.h [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+#include <cuda/__memory_resource/resource.h>
+
+#if _CCCL_STD_VER >= 2014
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA_MR
+
+/**
+ * @brief `cuda_pinned_memory_resource` uses cudaMallocHost / cudaFreeHost with cudaHostAllocPortable for allocation/deallocation.
+ */
+struct cuda_pinned_memory_resource
```
Comment (miscco) [inline libcudacxx/include/cuda/__memory_resource/cuda_pinned_memory_resource.h:40]: Should the `cuda_pinned_memory_resource` take passed in flags as an argument and hold that state?

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+{
+  /**
+   * @brief Allocate device memory of size at least \p __bytes.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation. Is ignored!
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* allocate(const size_t __bytes, const size_t __alignment) const
+  {
+    _LIBCUDACXX_ASSERT(__alignment <= 256 && (256 % __alignment == 0),
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:50]: suggestion: I think allocation should throw, not assert. Also note that this assertion/error means that the alignment parameter is NOT ignored.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+/**
+ * @brief `cuda_memory_resource` uses cudaMalloc / cudaFree for allocation/deallocation.
+ */
+struct cuda_memory_resource
+{
+  /**
+   * @brief Allocate device memory of size at least \p __bytes.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation. Is ignored!
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:45]: suggestion: this parameter is not ignored. And the requirements on it should be documented.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 5 above)
Comment (harrism) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:50]: Question: Just because cudaMalloc aligns to 256 (or possibly 512, even though documented at 256), doesn't mean an MR can't support larger allocations. I suppose users who *really* want that can write a wrapping resource that aligns, but is there another reason we don't support it here?  Seems a bit strange that cuda::mr has this parameter and our most basic implementation semi-ignores it and adds another method without it that is not part of the concept.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+  {
+    _LIBCUDACXX_ASSERT(__alignment <= 256 && (256 % __alignment == 0),
+                       "cuda_memory_resource::allocate invalid alignment");
+    return allocate(__bytes);
+  }
+
+  /**
+   * @brief Allocate device memory of size at least \p __bytes.
+   * @param __bytes The size in bytes of the allocation.
+   * @return void* Pointer to the newly allocated memory
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:58]: Suggestion: Document possible exception with @throw.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        _LIBCUDACXX_UNREACHABLE();
+#  endif
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The alignment that was passed to the `allocate` call that returned \p __ptr. Is ignored!
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:82]: Suggestion: Again, the alignment is not ignored. And there are requirements put on  that should be documented.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated
+   * @param __bytes The size in bytes of the allocation.
+   */
+  void deallocate(void* __ptr, size_t) const
+  {
+    const ::cudaError_t __status = ::cudaFree(__ptr);
+    (void)__status;
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:99]: question: Can we not use `[[maybe_unused]]`?

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+  {
+    _LIBCUDACXX_ASSERT(__alignment <= 256 && (256 % __alignment == 0),
+                       "cuda_memory_resource::deallocate invalid alignment");
+    deallocate(__ptr, __bytes);
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p __ptr.
+   * @param __ptr Pointer to be deallocated
+   * @param __bytes The size in bytes of the allocation.
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:94]: Question: should we document that the pointer and size should match the corresponding allocation? UB if a pointer is passed to deallocate that was not received from allocate of the same type of MR?

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+   */
+  void deallocate(void* __ptr, size_t) const
+  {
+    const ::cudaError_t __status = ::cudaFree(__ptr);
+    (void)__status;
+    _LIBCUDACXX_ASSERT(__status == cudaSuccess, "cuda_memory_resource::deallocate failed");
+  }
+
+  /**
+   * @brief Equality comparison operator between two cuda_memory_resource's
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:104]: suggestion: Apostrophe for possessives and contractions, not for plural. It's awkward when you want doxygen linking, but perhaps just simplify to use "another" rather than "two", and no need to say this is an operator. :)

```suggestion
   * @brief Equality comparison with another cuda_memory_resource
```

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+   * @return true
+   */
+  _LIBCUDACXX_NODISCARD_ATTRIBUTE constexpr bool operator==(cuda_memory_resource const&) const noexcept
+  {
+    return true;
+  }
+#  if _CCCL_STD_VER <= 2017
+  /**
+   * @brief Inequality comparison operator between two cuda_memory_resource's
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:113]:
```suggestion
   * @brief Inequality comparison with another cuda_memory_resource
```

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+   * @return false
+   */
+  _LIBCUDACXX_NODISCARD_ATTRIBUTE constexpr bool operator!=(cuda_memory_resource const&) const noexcept
+  {
+    return false;
+  }
+#  endif // _CCCL_STD_VER <= 2017
+
+  /**
+   * @brief Equality comparison operator between a cuda_memory_resource and a device_accessible resource
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:123]:
```suggestion
   * @brief Equality comparison between a cuda_memory_resource and a device_accessible resource
```

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+   *
+   * @param __lhs The cuda_memory_resource
+   * @param __rhs The resource to compare to
+   * @return false
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:127]: problem: does not return false.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 13 above)
Comment (harrism) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:123]: Suggestion: Perhaps a bit more explanation of the semantics are needed here. When is a cuda_memory_resource equal to any other device_accessible resource, and when is it not? Is there a section of the `memory_resource` docs that explains equality comparison that you can link to from these docs? If not, might be helpful.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+   */
+  _LIBCUDACXX_TEMPLATE(class _Resource)
+  _LIBCUDACXX_REQUIRES((!_CUDA_VSTD::same_as<_Resource, cuda_memory_resource>)
+                         _LIBCUDACXX_AND resource<_Resource> _LIBCUDACXX_AND has_property<_Resource, device_accessible>)
+  _LIBCUDACXX_NODISCARD_FRIEND bool operator==(cuda_memory_resource const& __lhs, _Resource const& __rhs) noexcept
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:132]: Question: Do these need docs? Judicious use of @copydoc, maybe?

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+                         _LIBCUDACXX_AND resource<_Resource> _LIBCUDACXX_AND has_property<_Resource, device_accessible>)
+  _LIBCUDACXX_NODISCARD_FRIEND bool operator!=(_Resource const& __rhs, cuda_memory_resource const& __lhs) noexcept
+  {
+    return resource_ref<device_accessible>{const_cast<cuda_memory_resource&>(__lhs)}
+        != resource_ref<device_accessible>{const_cast<_Resource&>(__rhs)};
+  }
+#  endif // _CCCL_STD_VER <= 2017
+
+  /**
+   * @brief Equality comparison operator between a cuda_memory_resource and an arbitrary resource
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:165]:
```suggestion
   * @brief Equality comparison between a cuda_memory_resource and an arbitrary resource
```

File: libcudacxx/include/cuda/memory_resource [hunk 17]
```diff
@@ -80,559 +80,22 @@ class resource_ref {
 */
 // clang-format on
 
-#ifdef LIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE
```
Comment (harrism) [inline libcudacxx/include/cuda/memory_resource:83]: Suggestion: I really think the refactoring in this PR should be in a separate PR from the new features added (cuda_memory_resource).

File: libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_memory_resource/allocate.pass.cpp [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+  { // allocate / deallocate
+    auto* ptr = res.allocate(42);
+    static_assert(cuda::std::is_same<decltype(ptr), void*>::value, "");
+    ensure_device_ptr(ptr);
+
+    res.deallocate(ptr, 42);
+  }
+
+  { // allocate / deallocate with alignment
+    auto* ptr = res.allocate(42, 4);
```
Comment (harrism) [inline libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_memory_resource/allocate.pass.cpp:39]: Suggestion: check that the returned pointer satisfies requested alignment.
Suggestion: test unsupported alignment throws (if you change assert to throw)?

**Review by harrism — CHANGES_REQUESTED** [review]
Nice additiojn. I didn't review the refactoring, only the new `cuda_memory_resource`. I think the refactoring is just splitting into multiple files, and that should be clarified in the PR description. It would have been nice to separate these into separate PRs.

Most of my suggestions are doc clarifications and fixes.

**Comment by harrism** [comment]
Also the title of this PR says "add ... memory resources" but I think only one is added?

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 13 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:123]: I added a larger explanation to why we compare the way we compare

This also made me realize that we should compare against a `resource_ref<>` as that is the lowest common denominator of all memory_resource and could also contain a `cuda_memory_resource`

Thanks a lot for bringing this up

File: libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_memory_resource/allocate.pass.cpp (hunk 18 above)
Comment (miscco) [reply libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_memory_resource/allocate.pass.cpp:39]: The returned pointer wont satisfy the requested alignment unless the default alignment was passed in. 

So I would rather not test cudaMalloc

I added two tests to see what happens if we throw exceptions. But I am not aware of any configuration that actually uses exceptions

Comment (jrhemstad) [reply libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_memory_resource/allocate.pass.cpp]: > The returned pointer wont satisfy the requested alignment unless the default alignment was passed in.

It _better_ satisfy the requested alignment! That's the whole point of the parameter!

To be clear, it's completely fine if the pointer has a _larger_ alignment than what was requested. The check should be for `actual_alignment >= requested_alignment`.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 5 above)
Comment (jrhemstad) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h]: > Seems a bit strange that cuda::mr has this parameter and our most basic implementation semi-ignores it and adds another method without it that is not part of the concept.

The existence of the alignment parameter does not mean all possible alignment values are valid. Each resource implementation will have its own limitations and this includes alignment. 

`cuda_memory_resource` is intended to be the simplest possible wrapper on top of `cudaMalloc`, and so it will inherit all the limitations of `cudaMalloc`, including alignment. 

We _could_ add an additional resource that allows for larger alignments, (or an adapter!), but that would no longer be a trivial wrapper on top of `cudaMalloc`.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+   * @param __ptr Pointer to be deallocated. Must have been allocated through a call to `allocate`
+   * @param __bytes The number of bytes that was passed to the `allocate` call that returned \p __ptr.
+   * @param __alignment The alignment that was passed to the `allocate` call that returned \p __ptr.
+   */
+  void deallocate(void* __ptr, const size_t __bytes, const size_t __alignment = __default_cuda_malloc_alignment) const
+  {
+    // We need to ensure that the provided alignment matches the minimal provided alignment
+    _LIBCUDACXX_ASSERT(
+      __default_cuda_malloc_alignment <= __alignment && (__alignment % __default_cuda_malloc_alignment == 0),
+      "Invalid alignment passed to cuda_memory_resource::deallocate.");
```
Comment (jrhemstad) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:88]: suggestion: I think this check is unnecessary. It's illegal to ever call `deallocate(p, N, M)` for `p` that was not returned by `p = allocate(N, M)`. Therefore, we would have already checked the alignment during the `allocate` call.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+   * @brief Allocate device memory of size at least \p __bytes.
+   * @param __bytes The size in bytes of the allocation.
+   * @param __alignment The requested alignment of the allocation.
+   * @throw cuda::cuda_error of the returned error code
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* allocate(const size_t __bytes, const size_t __alignment = __default_cuda_malloc_alignment) const
+  {
+    // We need to ensure that the provided alignment matches the minimal provided alignment
+    if (__alignment <= __default_cuda_malloc_alignment || (__alignment % __default_cuda_malloc_alignment != 0))
```
Comment (jrhemstad) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:52]: Looks like the logic was flipped here. 
```suggestion
    if (__alignment > __default_cuda_malloc_alignment || (__default_cuda_malloc_alignment % __alignment != 0))
```

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+    {
+#  ifndef _LIBCUDACXX_NO_EXCEPTIONS
+      throw ::cuda::cuda_error{cudaErrorInvalidValue, "Invalid alignment passed to cuda_memory_resource::allocate."};
```
Comment (jrhemstad) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:55]: `cuda_error` is really only intended for things returned from CUDA Runtime API calls, which is why it includes the `cudaError_t`. I don't think we want to get in the habit of using this for other things. This should be a different exception type. This could just be a `cuda::std::bad_alloc` for instance.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 20 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:52]: The logic is intentionally flipped here as inside the if we are throwing.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 19 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:88]: There is no guarantee that a user matches allocate and deallocate correctly. 

This is just a tiny bit safer than without and it does not hurt

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 20 above)
Comment (jrhemstad) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:52]: Right, I think the logic is still wrong. We want to throw if the alignment is invalid. The alignment is invalid if it is larger than what cudaMalloc supports or is not a power of 2.

The current logic does the opposite.

File: libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_memory_resource/allocate.pass.cpp (hunk 18 above)
Comment (harrism) [reply libcudacxx/test/libcudacxx/cuda/memory_resource/cuda_memory_resource/allocate.pass.cpp]: > It better satisfy the requested alignment! That's the whole point of the parameter!

Begging the question?

> The check should be for `actual_alignment >= requested_alignment`.

Yep, that's what I'm suggesting.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+  }
+#  endif // _CCCL_STD_VER <= 2017
+
+  /**
+   * @brief Equality comparison between a cuda_memory_resource and another resource
+   *
+   * We want to ensure that we can compare any cuda_memory_resource too a resource_ref that holds a cuda_memory_resource
+   * Therefore, we convert both to a resource_ref<> which is the lowest common denominator. The resource_ref<>
+   * internally checks whether both resources are of the same kind and if so returns the comparison of the two.
+   * Otherwise, the comparison will always return false, which is what we want for any other memory resource.
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:119]: Suggestion: This reads kind of like documentation of implementation details, especially with the "royal we".  Just document the semantics, which I think can be done in the @return statement.

```suggestion
```

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+   *
+   * @param __lhs The cuda_memory_resource
+   * @param __rhs The resource to compare to
+   * @return Comparison of both resources converted to a resource_ref<>
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:123]: Suggestion: The semantics are pretty simple, can we just spell it out something like the following (or whatever the incantation would be)?

```suggestion
   * @return true if both arguments are convertible to resource_ref<>(cuda_memory_resource&), false otherwise
```

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+   */
+  template <class _Resource>
+  _LIBCUDACXX_NODISCARD_FRIEND auto operator==(cuda_memory_resource const& __lhs, _Resource const& __rhs) noexcept
+    _LIBCUDACXX_TRAILING_REQUIRES(bool)(__different_resource<cuda_memory_resource, _Resource>)
+  {
+    return resource_ref<>{const_cast<cuda_memory_resource&>(__lhs)} == resource_ref<>{const_cast<_Resource&>(__rhs)};
+  }
+#  if _CCCL_STD_VER <= 2017
+  /**
+   * @copydoc cuda_memory_resource::operator==(cuda_memory_resource const&, _Resource const&)
```
Comment (harrism) [inline libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:133]: Question: did you verify that this works ? I recall copydoc requiring the exact prototype (e.g. template params may be necessary)

**Comment by harrism** [comment]
Typo in issue title (hyphen should be underscore).

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 21 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:55]: Urgh, I am really not looking forward into adding more error types, especially given that our current exception support is icky at best

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 24 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:133]: Yeah, I checked again and it seems I need to add `<_Resource>`

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 23 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:123]: I went with: Result of equality comparison of both resources converted to a resource_ref<>

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 21 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h:55]: I would like to postpone this, especially given that we currently do not expose `bad_alloc` and that is a lot of work

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 23 above)
Comment (harrism) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h]: OK, but then the reader has to go dig up docs on the semantics of comparison of refs. Doesn't that just compare the referenced object types? Is there a section in the docs on equality comparison that you could link to?

std::pmr::memory_resource and RMM (current) memory resources have very different comparison semantics, so I think this should be clearly documented.

File: libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h (hunk 21 above)
Comment (miscco) [reply libcudacxx/include/cuda/__memory_resource/cuda_memory_resource.h]: Moved to bad_alloc now that it is merged
